File: electron/utilis/tokenManager.ts
import Store from "electron-store";
import * as crypto from "crypto";

class SecureTokenManager {
  private store: Store;
  private readonly encryptionKey: Buffer;
  private readonly IV_LENGTH: number = 16;
  private readonly SALT_LENGTH: number = 16;
  private readonly ENCRYPTION_KEY_LENGTH: number = 32;

  constructor(masterKey: string = "your-secret-master-key") {
    // Generate a consistent encryption key from the master key
    const salt = crypto.scryptSync(
      masterKey,
      "fixed-salt",
      this.ENCRYPTION_KEY_LENGTH
    );
    this.encryptionKey = salt;

    // Initialize electron-store
    this.store = new Store({
      name: "secure-tokens", // Name of the storage file
      encryptionKey: masterKey, // Built-in encryption by electron-store
    });
  }

  private encrypt(text: string): string {
    // Generate a random IV for each encryption
    const iv = crypto.randomBytes(this.IV_LENGTH);

    // Create cipher with AES-256-GCM
    const cipher = crypto.createCipheriv("aes-256-gcm", this.encryptionKey, iv);

    // Encrypt the data
    let encrypted = cipher.update(text, "utf8", "hex");
    encrypted += cipher.final("hex");

    // Get authentication tag
    const authTag = cipher.getAuthTag();

    // Combine IV, encrypted data, and auth tag
    return `${iv.toString("hex")}:${encrypted}:${authTag.toString("hex")}`;
  }

  private decrypt(encryptedData: string): string {
    try {
      // Split the stored data into IV, encrypted text, and auth tag
      const [ivHex, encryptedText, authTagHex] = encryptedData.split(":");

      if (!ivHex || !encryptedText || !authTagHex) {
        throw new Error("Invalid encrypted data format");
      }

      const iv = Buffer.from(ivHex, "hex");
      const authTag = Buffer.from(authTagHex, "hex");

      // Create decipher
      const decipher = crypto.createDecipheriv(
        "aes-256-gcm",
        this.encryptionKey,
        iv
      );

      // Set auth tag for verification
      decipher.setAuthTag(authTag);

      // Decrypt the data
      let decrypted = decipher.update(encryptedText, "hex", "utf8");
      decrypted += decipher.final("utf8");

      return decrypted;
    } catch (error) {
      throw new Error("Failed to decrypt data: " + (error as Error).message);
    }
  }

  /**
   * Store a token securely
   * @param tokenType The type of token (e.g., 'id_token', 'refresh_token')
   * @param token The token value to store
   */
  storeToken(tokenType: string, token: string): void {
    try {
      const encryptedToken = this.encrypt(token);
      this.store.set(tokenType, encryptedToken);
    } catch (error) {
      console.error(`Error storing ${tokenType}:`, error);
      throw new Error(`Failed to store ${tokenType}`);
    }
  }

  /**
   * Retrieve a stored token
   * @param tokenType The type of token to retrieve
   * @returns The decrypted token or null if not found
   */
  retrieveToken(tokenType: string): string | null {
    try {
      const encryptedToken = this.store.get(tokenType) as string | undefined;

      if (!encryptedToken) {
        return null;
      }

      return this.decrypt(encryptedToken);
    } catch (error) {
      console.error(`Error retrieving ${tokenType}:`, error);
      throw new Error(`Failed to retrieve ${tokenType}`);
    }
  }

  /**
   * Delete a specific token
   * @param tokenType The type of token to delete
   */
  deleteToken(tokenType: string): void {
    try {
      this.store.delete(tokenType);
    } catch (error) {
      console.error(`Error deleting ${tokenType}:`, error);
      throw new Error(`Failed to delete ${tokenType}`);
    }
  }

  /**
   * Clear all stored tokens
   */
  clearAllTokens(): void {
    try {
      this.store.clear();
    } catch (error) {
      console.error("Error clearing tokens:", error);
      throw new Error("Failed to clear all tokens");
    }
  }

  /**
   * Check if a token exists
   * @param tokenType The type of token to check
   */
  hasToken(tokenType: string): boolean {
    return this.store.has(tokenType);
  }

  /**
   * Get all stored token types
   * @returns Array of token type strings
   */
  getStoredTokenTypes(): string[] {
    return Object.keys(this.store.store);
  }
}

export default SecureTokenManager;

File: backend/auth/token_manager.py
import os
import json
import base64
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
import appdirs


class SecureTokenManager:
    def __init__(self, master_key: str = "your-secret-master-key"):
        self.app_name = "pinac-workspace"
        self.data_dir = appdirs.user_data_dir(self.app_name)
        os.makedirs(self.data_dir, exist_ok=True)

        # Generate encryption key from master key
        key = self._derive_key(master_key)
        self.cipher = Fernet(key)

        self.token_file = os.path.join(self.data_dir, "secure-tokens.json")
        self._ensure_token_file()

    def _derive_key(self, master_key: str) -> bytes:
        """Derive encryption key from master key"""
        salt = b"fixed-salt-for-consistency"
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=salt,
            iterations=100000,
        )
        key = base64.urlsafe_b64encode(kdf.derive(master_key.encode()))
        return key

    def _ensure_token_file(self):
        """Ensure token file exists"""
        if not os.path.exists(self.token_file):
            with open(self.token_file, "w") as f:
                json.dump({}, f)

    def _load_tokens(self) -> dict:
        """Load encrypted tokens from file"""
        try:
            with open(self.token_file, "r") as f:
                return json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            return {}

    def _save_tokens(self, tokens: dict):
        """Save encrypted tokens to file"""
        with open(self.token_file, "w") as f:
            json.dump(tokens, f)

    def store_token(self, token_type: str, token: str):
        """Store a token securely"""
        try:
            encrypted_token = self.cipher.encrypt(token.encode()).decode()
            tokens = self._load_tokens()
            tokens[token_type] = encrypted_token
            self._save_tokens(tokens)
        except Exception as e:
            raise Exception(f"Failed to store {token_type}: {str(e)}")

    def retrieve_token(self, token_type: str) -> str | None:
        """Retrieve a stored token"""
        try:
            tokens = self._load_tokens()
            encrypted_token = tokens.get(token_type)

            if not encrypted_token:
                return None

            decrypted_token = self.cipher.decrypt(encrypted_token.encode()).decode()
            return decrypted_token
        except Exception as e:
            print(f"Error retrieving {token_type}: {str(e)}", flush=True)
            return None

    def delete_token(self, token_type: str):
        """Delete a specific token"""
        try:
            tokens = self._load_tokens()
            if token_type in tokens:
                del tokens[token_type]
                self._save_tokens(tokens)
        except Exception as e:
            raise Exception(f"Failed to delete {token_type}: {str(e)}")

    def clear_all_tokens(self):
        """Clear all stored tokens"""
        try:
            self._save_tokens({})
        except Exception as e:
            raise Exception(f"Failed to clear all tokens: {str(e)}")

    def has_token(self, token_type: str) -> bool:
        """Check if a token exists"""
        tokens = self._load_tokens()
        return token_type in tokens

    def get_stored_token_types(self) -> list:
        """Get all stored token types"""
        tokens = self._load_tokens()
        return list(tokens.keys())
